Mergecell speed

threeq opened the pull request:
# 提高 Merge Cell 执行速度

## Description

1. 修改 (*File).MergeCell 方法，放弃立即合并重叠单元格
2. 增加 (*File).mergeOverlapCells 方法，用以合并重叠单元格。在`(*File).UnmergeCell`、`(*File).GetMergeCells`、`(*File).writeToZip` 中调用。
3. 增加 xlsxMergeCell.rect 属性，避免重复计算合并单元格坐标

## Related Issue

#595 Poor performance when MergeCell and SetCellValue In a for loop 

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## 测试情况

* MacBook Pro (Retina, 13-inch, Early 2015)
* CPU 3.1 GHz 双核Intel Core i7
* Memory 8 GB 1867 MHz DDR3 

![image](https://user-images.githubusercontent.com/4645905/126884901-f6b25b35-8976-4e8f-9b0d-d4af4e1cb973.png)


## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] Docs change / refactoring / dependency upgrade

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

xuri commented on merge_test.go:
```diff
@@ -181,3 +191,60 @@ func TestUnmergeCell(t *testing.T) {
+	}
+	headerRow = 1
+
+	//问题的关键
```
Remove the unrelated comments of the function.

threeq replied:
It removed.

threeq replied:
Overlapping merge problem has been resolved.

xuri reviewed (changes requested):
Thanks for your PR. I've left some comments. Please don't commit the spreadsheet generated after the test, and add the unit test for the uncovered line of the code. This change will generate a corrupted file when setting two or more overlapped merge cells, for example: `A1:C2` and `B2:D3`, in addition, the final range of overlap merge cells range should be already known on insert the picture with autofit the cell size, so I think the merge overlap range progress shouldn't delay to saving the spreadsheet.

threeq commented:
我们现在遇到的问题是需要写入 10w 数据，里面有大量自动合并单元格的需求，发现系统直接蹦了。

![image](https://user-images.githubusercontent.com/4645905/127014389-8a6da2aa-f19f-4409-b91e-5a60e0929641.png)

以上数据是相同单元测试 `TestMergeCellSpeed` 在 `master` 分支跑的结果，2w数据写入我没有跑，这个得按照小时来算了。

![image](https://user-images.githubusercontent.com/4645905/127013347-e6b49806-f5e1-4947-9858-93ce320bc9fe.png)

从 CPU 耗时来看，大部分时间都耗费在了 `MergeCell` 和 `areaRefToCoordinates` 这两个方法，里面又有 50% 的时间在做 `strings.Split`。

分析原因在于每次增加一个 `MergeCell`，就会全部计算一遍覆盖，随着增加的数量越多。第 N 次增加时的计算量都是 `N-1`，整个程序的复杂度是 `N^2`。所以这里做了 2 个优化：

> 1. 缓存单个 MergeCell 的 Rect 计算记过，减少 `strings.Split` 调用
> 2. 最后只做一次计算，时间复杂度降低到 `N`

经过优化，目前耗时已经集中在 `writeToZip` 写磁盘的时候了，基本符合预期
![image](https://user-images.githubusercontent.com/4645905/127015536-eba3f232-9df9-4b08-9a21-01178b3a0c64.png)

xuri reviewed (changes requested):
Hi @threeq, thanks for your PR, this change caused the merged cell range calculation error, the spreadsheet `TestMergeCell.xlsx` generated by the unit test was corrupted.

threeq commented:
Hi @xuri，谢谢你的审核，问题已经修复。
是在计算单元格合并的时候没有考虑到大于 2 个覆盖的情况，算法已经修改。

xuri commented:
Thanks for your contribution. I have committed some changes base on your branch, in this commit:

* Separate mergeOverlapCells into `overlapRange` and `flatMergedCells` reduce cyclomatic complexities 
* Make some variables name semantic and reuse some internal coordinates helper function
* Fix merge cell adjust helper issue when removing the column in the left boundary range
* Move the structure implementation code of the `xlsxMergeCell` and `xlsxC` into each module source code file, to make a code file that has `xml` prefix only contains structure definition, note that these part files can be generated by the XML schema definition (XSD), and I'll consider also hold the unexported  field `rect` in another place
